### ISSUE TITLE
TypeError: getaddrinfo() argument 1 must be string or None

### ISSUE THREAD

**Issue by likewhoa** [aio-libs/aiohttp#4557]
🐞 **Describe the bug**
While trying to fetch urls I am getting the above error message and I recalled a while ago that it was caused by an ipv6 line in /etc/hosts so I checked to see if that had been uncommented out and that was not the case, so now I don't know what could be the issue. The server has no ipv6 support enabled or any ipv6 addresses.

💡 **To Reproduce**

You can reproduce by running the client example code at https://docs.aiohttp.org/en/stable/#getting-started

💡 **Expected behavior**
Fetch website html

📋 **Logs/tracebacks**

```python-traceback
Traceback (most recent call last):
  File "test.py", line 15, in <module>
    loop.run_until_complete(main())
  File "/usr/lib64/python3.6/asyncio/base_events.py", line 488, in run_until_complete
    return future.result()
  File "test.py", line 10, in main
    html = await fetch(session, 'http://python.org')
  File "test.py", line 5, in fetch
    async with session.get(url) as response:
  File "/usr/lib64/python3.6/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/usr/lib64/python3.6/site-packages/aiohttp/client.py", line 483, in _request
    timeout=real_timeout
  File "/usr/lib64/python3.6/site-packages/aiohttp/connector.py", line 523, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/lib64/python3.6/site-packages/aiohttp/connector.py", line 859, in _create_connection
    req, traces, timeout)
  File "/usr/lib64/python3.6/site-packages/aiohttp/connector.py", line 986, in _create_direct_connection
    req=req, client_error=client_error)
  File "/usr/lib64/python3.6/site-packages/aiohttp/connector.py", line 936, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/usr/lib64/python3.6/asyncio/base_events.py", line 754, in create_connection
    infos = f1.result()
  File "/usr/lib64/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/lib64/python3.6/socket.py", line 745, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
TypeError: getaddrinfo() argument 1 must be string or None
```

📋 **Your version of the Python**

```console
$ python --version
Python 3.6.10
...
```

📋 **Your version of the aiohttp/yarl/multidict distributions**
```console
$ python -m pip show aiohttp

Name: aiohttp
Version: 3.6.1
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: Nikolay Kim
Author-email: fafhrd91@gmail.com
License: Apache 2
Location: /usr/lib64/python3.6/site-packages
Requires: attrs, chardet, multidict, async-timeout, yarl, idna-ssl, typing-extensions
Required-by: discord.py
```
```console
$ python -m pip show yarl

Name: yarl
Version: 1.3.0
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/lib64/python3.6/site-packages
Requires: multidict, idna
Required-by: aiohttp
```

📋 **Additional context**

**Comment by socketpair** [comment]
@likewhoa I can not reproduce the issue. Please give exact steps, or at least debug on your system what is sent to getaddrinfo as the first argument.

Without that the issue can not be solved.

**Comment by likewhoa** [comment]
@socketpair I am basically reproducing with the example client code at https://docs.aiohttp.org/en/stable/#getting-started

**Comment by hh-h** [comment]
You forgot to specify version of aiohttp. Also, it can be related to your computer configuration, any chance that you can run it on another machine?

**Comment by webknjaz** [comment]
> You forgot to specify version of aiohttp

He specified it but posted under the wrong section (the command says `multidict` but in fact, the output is for `aiohttp`):
> ```console
> $ python -m pip show multidict
> ...
> 
> Name: aiohttp
> Version: 3.6.1
> Summary: Async http client/server framework (asyncio)
> Home-page: https://github.com/aio-libs/aiohttp
> Author: Nikolay Kim
> Author-email: fafhrd91@gmail.com
> License: Apache 2
> Location: /usr/lib64/python3.6/site-packages
> Requires: attrs, chardet, multidict, async-timeout, yarl, idna-ssl, typing-extensions
> Required-by: discord.py
> ```

**Comment by likewhoa** [comment]
for multidict info this is it
```
Name: multidict
Version: 4.5.2
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/lib64/python3.6/site-packages
Requires: 
Required-by: yarl, aiohttp
```

@hh-h i can't reproduce on another server

**Comment by webknjaz** [comment]
@likewhoa to make this report useful, we need a stable reproducer, I don't think we can move the investigation forward w/o a way to reproduce it on any dev's env. You need to find out what the difference is between your two envs where you try to run that code.

**Comment by likewhoa** [comment]
> @likewhoa to make this report useful, we need a stable reproducer, I don't think we can move the investigation forward w/o a way to reproduce it on any dev's env. You need to find out what the difference is between your two envs where you try to run that code.

What are some steps that you would suggest. I verified that aiohttp/yarl/multidict versions match on both servers. The other difference is kernel with the server with issues having kernel 5.0.x while on server were I was no able to reproduce issue is running 5.4.x kernel.

**Comment by webknjaz** [comment]
> ```python-traceback
>  File "/usr/lib64/python3.6/socket.py", line 745, in getaddrinfo
>    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
> ```

You could try to stick a debugger here and see what the args are and then trace it to the calls in your app or in aiohttp at least.

**Comment by likewhoa** [comment]
@webknjaz both servers are showing this for args
```python
host=python.org,port=80,family=0,type=1,proto=0,flags=0
host=python.org,port=443,family=0,type=1,proto=0,flags=0
host=www.python.org,port=443,family=0,type=1,proto=0,flags=0
```
but server in question has this extra one
```python
host=10,port=b'\x01\xbb\x00\x00\x00\x00*\x04NB\x00\x05\x00\x00',family=10,type=1,proto=6,flags=4
```
both using en_US.utf8 locale.

**Comment by webknjaz** [comment]
Well, this is weird. `b'\x01\xbb\x00\x00\x00\x00*\x04NB\x00\x05\x00\x00'` is not something UTF8-encoded. And also how did this get into port?

You need to go up the stack and grab the local var values from each frame up...

**Comment by likewhoa** [comment]
> Well, this is weird. `b'\x01\xbb\x00\x00\x00\x00*\x04NB\x00\x05\x00\x00'` is not something UTF8-encoded. And also how did this get into port?
> 
> You need to go up the stack and grab the local var values from each frame up...

I will do that, I did a ```diff``` from both server and the file doesn't differ. I will go ahead and debug further as suggested. Be back in a bit.

**Comment by webknjaz** [comment]
Also, try to check DNS records of `python.org`/`www.python.org` (is this what you're querying?) using a `dig` utility from both of your servers: are the A/AAAA records different?

**Comment by webknjaz** [comment]
> ```python
> family=10
> ```

Looks like it's an IPv6 address:
`socket.AF_INET6 == 10`

Also, calling that socket method gives me this:
```python
>>> socket.getaddrinfo('www.python.org', 80)                              

[(<AddressFamily.AF_INET: 2>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('151.101.112.223', 80)),
 (<AddressFamily.AF_INET: 2>,
  <SocketKind.SOCK_DGRAM: 2>,
  17,
  '',
  ('151.101.112.223', 80)),
 (<AddressFamily.AF_INET: 2>,
  <SocketKind.SOCK_RAW: 3>,
  0,
  '',
  ('151.101.112.223', 80)),
 (<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_STREAM: 1>,
  6,
  '',
  ('2a04:4e42:1b::223', 80, 0, 0)),
 (<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_DGRAM: 2>,
  17,
  '',
  ('2a04:4e42:1b::223', 80, 0, 0)),
 (<AddressFamily.AF_INET6: 10>,
  <SocketKind.SOCK_RAW: 3>,
  0,
  '',
  ('2a04:4e42:1b::223', 80, 0, 0))]
```

**Comment by webknjaz** [comment]
Looking at the traceback again, it appears that this broken value may be coming from https://github.com/aio-libs/aiohttp/blob/4a08f89/aiohttp/connector.py#L943-L956. Worth checking on the affected server with a debugger.

**Comment by likewhoa** [comment]
@webknjaz yea it's not just with python.org, the reason you're seeing that is because I pointed to the official client example. A while ago, i had same issue and commenting out the ipv6 line in /etc/hosts fixed it but this is not the case this time, perhaps it's looking else?

**Comment by webknjaz** [comment]
Do you use any other domains in your code?

**Comment by likewhoa** [comment]
@weastur looks like https://github.com/aio-libs/aiohttp/blob/4a08f89/aiohttp/connector.py#L943 is what is returning 
```
hosts = [{'hostname': 'google.com', 'host': 10, 'port': b'\x00P\x00\x00\x00\x00&\x07\xf8\xb0@\x06\x08\x19', 'family': <AddressFamily.AF_INET6: 10>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}, {'hostname': 'google.com', 'host': '172.217.10.142', 'port': 80, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}]
``` 
when changing the domain, so asyncio.shield is what is putting in that port. 

with regards to my code, it uses many domains but i am reproducing this with a simple example aiohttp client code found upstream.

**Comment by socketpair** [comment]
What is used -- aiodns (pycares) or thread resolver ? I bet, first.

**Comment by likewhoa** [comment]
> What is used -- aiodns (pycares) or thread resolver ? I bet, first.

aiodns not on the system. I am going to disable ipv6 all together on system though to rule that out

**Comment by JustinTArthur** [comment]
@likewhoa is on Gentoo using Portage-managed Python packages. I tried to reproduce with the following package versions:
* sys-libs/glibc-2.29-r2
* dev-lang/python-3.6.9
* dev-python/aiohttp-3.6.1
* dev-python/idna-ssl-1.1.0
* dev-python/multidict-4.5.2
* dev-python/yarl-1.3.0

but everything worked as normal. Because aiohttp, multidict and yarl all have compiled C code, one possibility is glibc was upgraded or the Gentoo profile was changed after one or more of them had been installed. @likewhoa is that possible? If so, can you try again after re-emerging aiohttp, multidict, and yarl?

**Comment by likewhoa** [comment]
@webknjaz @socketpair so disabling ipv6 fixes the issue but now makes you wonder why would ipv6 support enable be the issue.
```
sysctl -w net.ipv6.conf.all.disable_ipv6=1
sysctl -w net.ipv6.conf.default.disable_ipv6=1
```

**Comment by likewhoa** [comment]
> @likewhoa is on Gentoo using Portage-managed Python packages. I tried to reproduce with the following package versions:
> 
> * sys-libs/glibc-2.29-r2
> * dev-lang/python-3.6.9
> * dev-python/aiohttp-3.6.1
> * dev-python/idna-ssl-1.1.0
> * dev-python/multidict-4.5.2
> * dev-python/yarl-1.3.0
> 
> but everything worked as normal. Because aiohttp, multidict and yarl all have compiled C code, one possibility is glibc was upgraded or the Gentoo profile was changed after one or more of them had been installed. @likewhoa is that possible? If so, can you try again after re-emerging aiohttp, multidict, and yarl?

I didn't see there was any changes to the profile or packages relating to python, so I figured it was something else and my first suspect was ipv6 support and once i disabled that everything started working. We would need to figure out why this is the case though.

**Comment by JustinTArthur** [comment]
Strange. On that Gentoo testbed where I couldn't repro:
```
$ sudo sysctl -a | grep net.ipv6.conf.*disable
net.ipv6.conf.all.disable_ipv6 = 0
net.ipv6.conf.all.disable_policy = 0
net.ipv6.conf.default.disable_ipv6 = 0
net.ipv6.conf.default.disable_policy = 0
net.ipv6.conf.enp39s0.disable_ipv6 = 0
net.ipv6.conf.enp39s0.disable_policy = 0
```
I'd expect IPv6 come into play *during* the `getaddrinfo` and not before

**Comment by likewhoa** [comment]
@JustinTArthur strange indeed but since I don't have use for ipv6 it's best to be disabled until I actually need it on this server. On my workstation I also have ipv6 enabled and i can't reproduce.

**Comment by likewhoa** [comment]
> @likewhoa is on Gentoo using Portage-managed Python packages. I tried to reproduce with the following package versions:
> 
> * sys-libs/glibc-2.29-r2
> * dev-lang/python-3.6.9
> * dev-python/aiohttp-3.6.1
> * dev-python/idna-ssl-1.1.0
> * dev-python/multidict-4.5.2
> * dev-python/yarl-1.3.0
> 
> but everything worked as normal. Because aiohttp, multidict and yarl all have compiled C code, one possibility is glibc was upgraded or the Gentoo profile was changed after one or more of them had been installed. @likewhoa is that possible? If so, can you try again after re-emerging aiohttp, multidict, and yarl?

Here is my setup, the only difference is glibc and probably gcc as I use ~arch (unstable) but my workstation matches this and it doesn't reproduce the issue even with ipv6 enabled

> sys-libs/glibc-2.30-r3
> sys-devel/gcc-9.2.0-r3
> dev-lang/python-3.6.10
> dev-python/aiohttp-3.6.1
> dev-python/idna-ssl-1.1.0
> dev-python/multidict-4.5.2
> dev-python/yarl-1.3.0

and my profile is but keep in mind, i was using 17.0 when i ran into issue, i just updated to 17.1 just now.
```
# eselect profile show
Current /etc/portage/make.profile symlink:
  default/linux/amd64/17.1/no-multilib
```

**Comment by socketpair** [comment]
@likewhoa Please check if it is triggered in you environment:

```python
import socket

print(socket.getaddrinfo('ipv6.google.com', 80))
```

If it triggers the problem, the bug is definitely not in aiohttp. Please report to bugs.python.org. I would recommend to debug thoroughly before sending the bug report.

**Comment by socketpair** [comment]
@likewhoa

> 'port': b'\x00P\x00\x00\

First two bytes are `80` encoded as `unsigned short`.

> port=b'\x01\xbb\x00\

The same here, but port is `443`


https://github.com/aio-libs/aiohttp/blob/v3.6.1/aiohttp/connector.py#L774

Try disabling dns cache. and re-check.

```python
conn = aiohttp.TCPConnector(use_dns_cache=False)
session = aiohttp.ClientSession(connector=conn)
```

**Comment by socketpair** [comment]
https://github.com/aio-libs/aiohttp/blob/v3.6.1/aiohttp/connector.py#L774

Try disabling dns cache. and re-check.

**Comment by likewhoa** [comment]
> @likewhoa
> 
> > 'port': b'\x00P\x00\x00\
> 
> First two bytes are `80` encoded as `unsigned short`.
> 
> > port=b'\x01\xbb\x00\
> 
> The same here, but port is `443`
> 
> https://github.com/aio-libs/aiohttp/blob/v3.6.1/aiohttp/connector.py#L774
> 
> Try disabling dns cache. and re-check.
> 
> ```python
> conn = aiohttp.TCPConnector(use_dns_cache=False)
> session = aiohttp.ClientSession(connector=conn)
> ```

I will try this later today and get back to you with results.

**Comment by socketpair** [comment]
@likewhoa I'm very curious to ask how is it going ? :)

**Comment by likewhoa** [comment]
> @likewhoa I'm very curious to ask how is it going ? :)

sorry got distracted since I got the workaround going but I will try your suggestion this coming week.

**Comment by likewhoa** [comment]
> @likewhoa
> 
> > 'port': b'\x00P\x00\x00\
> 
> First two bytes are `80` encoded as `unsigned short`.
> 
> > port=b'\x01\xbb\x00\
> 
> The same here, but port is `443`
> 
> https://github.com/aio-libs/aiohttp/blob/v3.6.1/aiohttp/connector.py#L774
> 
> Try disabling dns cache. and re-check.
> 
> ```python
> conn = aiohttp.TCPConnector(use_dns_cache=False)
> session = aiohttp.ClientSession(connector=conn)
> ```

No go, here is the sample code I used to test 
```python
import aiohttp
import asyncio

async def fetch(session, url):
    async with session.get(url) as response:
        return await response.text()

async def main():
    conn = aiohttp.TCPConnector(use_dns_cache=False)

    async with aiohttp.ClientSession(connector=conn) as session:
        html = await fetch(session, 'http://python.org')
        print(html)

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main()
```
Can you confirm the glibc version you tested with? I am using sys-libs/glibc-2.30-r3 and someone else suggested this could be the issue but I ran into this once with a previous glibc version so I am assuming that's not the real cause here. At this point I just disabled ipv6 with sysctl to work around it.

**Comment by socketpair** [comment]
Try to reduce down the code even more. For example, try to use just the connector.

I mean
```python
import aiohttp
import asyncio
from yarl import URL
from aiohttp.client import ClientRequest, DEFAULT_TIMEOUT

async def main():
    conn = aiohttp.TCPConnector(use_dns_cache=False)
    rq = ClientRequest('GET', URL('http://python.org'))
    c = await conn.connect(rq, [], DEFAULT_TIMEOUT)

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

**Comment by socketpair** [comment]
If it still fails, try copy-pasting the `connect` method and reduce it

**Comment by likewhoa** [comment]
@socketpair 
```
Unclosed connection
client_connection: Connection<ConnectionKey(host='python.org', port=80, is_ssl=False, ssl=None, proxy=None, proxy_auth=None, proxy_headers_hash=None)>
```
output

**Comment by danielnelson** [comment]
I ran into a very similar issue today with aiohttp==3.6.3 and Python 3.7.9.

For me, it appears to the difference between calling `getaddrinfo` with 0 vs AF_INET.  I don't compile my kernel with IPv6 support, possibly this is the cause.

```python
$ cat bug.py 
import asyncio
import socket
from aiohttp.resolver import ThreadedResolver


async def main():
    resolver = ThreadedResolver()
    print("0:", await resolver.resolve('localhost', 8080, family=0))
    print(socket.getaddrinfo('localhost', 8080, type=socket.SOCK_STREAM, family=0))
    print("===")
    print("AF_INET:", await resolver.resolve('localhost', 8080, family=socket.AF_INET))
    print(socket.getaddrinfo('localhost', 8080, type=socket.SOCK_STREAM, family=socket.AF_INET))


asyncio.run(main())
```

```python
$ python bug.py
0: [{'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}, {'hostname': 'localhost', 'host': 10, 'port': b'\x1f\x90\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 'family': <AddressFamily.AF_INET6: 10>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}]
[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', (10, b'\x1f\x90\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'))]
===
AF_INET: [{'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}, {'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}]
[(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080))]
```

**Comment by likewhoa** [comment]
> I ran into a very similar issue today with aiohttp==3.6.3 and Python 3.7.9.
> 
> For me, it appears to the difference between calling `getaddrinfo` with 0 vs AF_INET. I don't compile my kernel with IPv6 support, possibly this is the cause.
> 
> ```python
> $ cat bug.py 
> import asyncio
> import socket
> from aiohttp.resolver import ThreadedResolver
> 
> 
> async def main():
>     resolver = ThreadedResolver()
>     print("0:", await resolver.resolve('localhost', 8080, family=0))
>     print(socket.getaddrinfo('localhost', 8080, type=socket.SOCK_STREAM, family=0))
>     print("===")
>     print("AF_INET:", await resolver.resolve('localhost', 8080, family=socket.AF_INET))
>     print(socket.getaddrinfo('localhost', 8080, type=socket.SOCK_STREAM, family=socket.AF_INET))
> 
> 
> asyncio.run(main())
> ```
> 
> ```python
> $ python bug.py
> 0: [{'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}, {'hostname': 'localhost', 'host': 10, 'port': b'\x1f\x90\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 'family': <AddressFamily.AF_INET6: 10>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}]
> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', (10, b'\x1f\x90\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'))]
> ===
> AF_INET: [{'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}, {'hostname': 'localhost', 'host': '127.0.0.1', 'port': 8080, 'family': <AddressFamily.AF_INET: 2>, 'proto': 6, 'flags': <AddressInfo.AI_NUMERICHOST: 4>}]
> [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 8080))]
> ```

same here, I don't compile ipv6 since I am not using it.

**Comment by Dreamsorcerer** [comment]
I think this may be related to issues in aiodns which have been fixed. Can anyone try reproducing it again?

**Comment by bdraco** [comment]
likely fixed by #8295 or a cpython change

**Comment by bdraco** [comment]
No response, and cannot reproduce. I'm going to assume this is fixed by https://github.com/aio-libs/aiohttp/pull/8295
